### ISSUE TITLE
There is no way to exit a text field in a web application with NVDA

### ISSUE THREAD

**Issue by titet11** [nvaccess/nvda#15619]
When you enter the textfree.us page, NVDA positions itself in the text box to send a message.

The problem is that I can't get out of this text field.

There are many pages that contain these edit fields, but I don't remember them at the moment.

**Comment by seanbudd** [comment]
Welcome @titet11 ,

We are unable to process this issue as it stands.

- Please open a new issue and fill out the bug report template fully.

- This is very likely a problem with the implementation of the application/website, as a starting point please submit feedback to the authors to let them know about the problem.

Kind Regards,
NV Access Software Developers

**Comment by titet11** [comment]
@seanbudd 
No, this is not a website problem.

I have used this site for years and it has always been accessible.

The only way to get out of this edit box is to move the mouse cursor.

**Comment by seanbudd** [comment]
Websites can update their code and implementation.
Again, please report this to the web authors and open a new issue with the required information to investigate.

**Comment by titet11** [comment]
@seanbudd 
I have a question.

Why is your goal for websites to adapt their code to screen readers, rather than NVDA trying to adapt its own code to inaccessible websites?

Or is it for budget reasons?

You already know that if you report problems to a website, they almost never pay attention to minor problems, unless they are security problems.

**Comment by seanbudd** [comment]
It is up to web authors to adhere to accessibility standards. We are unable to handle cases where there are development errors.

Just as a developer needs to make code that is bug free for sighted users, it also requires not introducing bugs for screen reader users.

From a technical perspective, code must be fixed by the authors introducing the error.

Handling web author issues create contradictions, i.e. creating special handling for an authoring issue will introduce bugs for people who follow standards correctly.

There are also budget constraints you mention - even if no contradictions would be introduced, screen reader developers do not have the capacity to handle special cases for the millions of websites that exist.

**Comment by seanbudd** [comment]
It is up to web authors to adhere to accessibility standards. We are unable to handle cases where there are development errors.

Just as a developer needs to make code that is bug free for sighted users, it also requires not introducing bugs for screen reader users.

From a technical perspective, code must be fixed by the authors introducing the error.

Handling web author issues create contradictions, i.e. creating special handling for an authoring issue will introduce bugs for people who follow standards correctly.

There are also budget constraints you mention - even if no contradictions would be introduced, screen reader developers do not have the capacity to handle special cases for the millions of websites that exist.

**Comment by josephsl** [comment]
Hi,

You may need to create an account on textfree.us to test the issue at hand.

Some comments:

1. Please follow the steps asked by NV Access and other NVDA developers. Part of what makes this issue a bit harder to diagnose is that we cannot imagine what is going on as there is no steps to reproduce the accessibility problem stated.
2. To verify that this is not a website issue, let us test with multiple screen readers and various browsers to make sure the issue can be pinpointed to a screen reader/browser combination.
3. I'm confused with the statement that the website was always accessible while being used for years. To put it the other way, the claim that something is suddenly inaccessible or unusable (note these two different terms) without giving people how to reproduce the issue does not make sense. Is NVDA saying "application" when opening this website and the edit field is focused? Is this happening with older NVDA versions and/or with other screen readers? Which web browser is in use? Lack of info like this has changed how the issue is communicated and understood by NVDA developers, so please keep this in mind.

Side note: I intentionally played dumb in item 3 to use this situation as a teachable moment - to help folks understand the effect of the miscommunication playing out at the moment.

Hope this helps. Thanks.

**Comment by titet11** [comment]
@josephsl 

Hello: I confirm that I have just tried other screen readers such as Jaws and Windows Elles, and it is basically the same problem.

2. I'll try to be more specific.

When I log in to textfree.us, NVDA is automatically placed in the edit box to send an SMS message.

However, the problem is that I can't get out of this edit box unless I move the mouse cursor or press the right mouse key, which basically opens the context menu.
3. The edit box that texfree.us has is no longer the same as before, so there have been recent changes. This new edit box is based on the web applications they are releasing recently.
If I'm not mistaken, these architectures are called web applications, better known as web apps, if I'm not mistaken...

When NVDA enters this type of edit box, NVDA announces that it is a web application.
This same architecture is the same one that WhatsApp implemented on October 22 of last year. This architecture is problematic for screen readers, as it prevents messages from being read by characters, lines, or paragraphs.

**Comment by josephsl** [comment]
Hi, at least that tells us what’s going on, and sounds like this is indeed something going on with the web application. In this case, try switching to browse mode (NVDA+Space) and see if that makes a difference. Thanks.

**Comment by titet11** [comment]
@josephsl 
Unfortunately that didn't work for me.

What I would like is some way that the Escape key or arrow keys can exit that web application that contains the edit box.

I would also like NVDA to be able to re-read Instagram messages, and I assume they would have to carefully analyze Instagram's source code to do so.

We must remember that Instagram is one of the most used sites and therefore, it is necessary to work on it.

I have already reported this problem to Instagram but they care little or nothing about solving it.

**Comment by ABuffEr** [comment]
Hi @titet11,
you probably need to use NVDA+control+space, much less known of NVDA+space (but anyway documented in command list).
With the growth of application in web trends, I could agree that it could be mapped (also) to escape key, but, nevertheless and again, take care in filling the template when you open an issue, it's the basic starting point to have a productive discussion, even when the question seems apparently obvious.

**Comment by titet11** [comment]
@ABuffEr 
You surprised me, brother.

This really solved my problem.

but why use this keyboard shortcut when you could attribute it to the Escape key?

The keys I pressed were:

Insert + control + space

I hope that in future versions this keyboard shortcut can be attributed to the Escape key.

I also hope they find some way to make NVDA read Instagram messages, as this is too necessary for many.

**Comment by josephsl** [comment]
Hi,

The reason is that Escape key can be used by the web application for various things (for example, it can be used in a web-based email application to close a message, and in case you are drafting a message, prompt if you would like to save or abandon the draft).

Also, one more advice: please do not hijack a GitHub issue to talk about something else entirely unless the underlying issue being discussed touches many things at once (directly or indirectly).

Thanks.

**Comment by titet11** [comment]
@josephsl 
Let's be honest, most of us use the Escape key to exit a form and not to cancel a specific dialog.

In old versions of NVDA, I remember that to exit a form was by pressing ALT + up arrow keys.

That shortcut from the past was more convenient than pressing insert + control + space.

Since this shortcut is very useful for exiting a form, wouldn't it be more convenient to add an option to be able to customize this shortcut manually from the keyboard options? I don't think it will be difficult to implement this functionality. I hope they implement it as soon as possible.

**Comment by ruifontes** [comment]
Go to NVDA menu, Preferences, Input gestures...
Look for System focus, open the item, look for "Moves the focus out of the current embedded object and into the document that contains it" and assign the keystroke you want...

**Comment by josephsl** [comment]
Hi,

Do note that Alt+Up arrow is used to close an expandable element such as a combo box.

Escape key is used to close something, and in some cases, it can be understood as canceling an operation (if you press Escape key while copying a large folder or two via File Explorer, Windows will say "cancelling" instead of "exiting").

What Control+NVDA+Space does is it moves the browse mode cursor to the "parent" web document. While it is phrased a bit oddly in NVDA's user interface, the idea is that a web application is "contained" (or shown) inside a web document. This command, originally used to leave Flash content and other embedded objects on websites, is also used to "close" a web application interface by telling NVDA to move browse mode cursor to a sensible location on the parent document (remember: browse mode is a "linear" representation of a web document, which is actually organized in a tree structure). When we claim to use browse mode to interact with documents and web apps, we are essentially saying that we use NVDA's linear "presentation" of web content as "lens" by which we interact with the document and ap in question.

CC @JCSTeh if you have something to say on this matter.

**Comment by titet11** [comment]
@josephsl 
Wouldn't it be more convenient if the down arrow was responsible for expanding and collapsing, instead of the up arrow being responsible for collapsing elements?

Currently the down arrow can already perform this action without problems, so it seems unnecessary to use the up arrow to collapse.

Additionally, when a combo box is expanded, NVDA already has the functionality to describe when an item is collapsed and expanded.

**Comment by josephsl** [comment]
Hi,

As explained in #8541, Alt+up/down arrow keys in combo boxes is a Windows command to collapse/expand combo boxes.

Thanks.

**Comment by titet11** [comment]
@josephsl 
I think you couldn't understand me:

What I meant was that the down arrow can perform both actions such as contracting and expanding, so using the up arrow is totally unnecessary.

**Comment by josephsl** [comment]
Hi,

True. But do note that when I tested with Narrator, Alt+up arrow also expanded and collapsed combo boxes, meaning that in this case, Alt+up arrow is a command handled by Windows (more specifically, by the GUI).

Thanks.

**Comment by titet11** [comment]
@josephsl It matters little or nothing that it is managed by Windows.

In previous versions, the alt up arrow was used to exit a form, i.e. basically the same shortcut as insert + control + space.

**Comment by josephsl** [comment]
Hi,

Perhaps the command has changed in order to acknowledge the fact that Alt+up arrow is used in places other than browse mode for various purposes. Also, think about web apps that could dedicate Alt+up arrow keys to perform their things. At least adding NVDA key to the mix to move to the parent document is also an acknowledgement that the command is NVDA specific (not all NVDA commands use NVDA key as part of their key assignments, however). Does Narrator use Alt+up arrow keys to exit out of web apps, and how do other screen readers work in this regard? If there is a case for Alt+up arrow to become the command to move to parent document in browse mode, perhaps folks might consider that, although you will come across comments like the one I wrote about commands and such (and that's reality of a public repo like this).

If I'm reading the comments right, I get an impression that there seems to be a wide valley between us (perhaps I might not be getting the full picture, and if that's the case, I apologize). I hope that this valley can narrow as time goes.

Hope this helps. Thanks.

**Comment by XLTechie** [comment]
@titet11
This is now the second of your issues (that I know of), where you decided not 
to use the template, despite many requests to do so.

I wonder if you do not know where the template is, or what this means?

If you go to https://github.com/nvaccess/nvda/issues, and select "new issue", 
there are three choices for issue type. Mostly, you will want to use the get 
started link for the bug report, or the feature request.

When you open one of those issues, the textarea where you enter your comment, 
already contains some text, with several questions that you are supposed to 
answer. So, when you enter this text area, don't just start typing, but actually 
use your arrow keys to examine and type answers to what is already there.

If you do not do this, your issues will continue to be misunderstood and closed.

**Comment by titet11** [comment]
@josephsl 
So far, there is no other application that uses Alt plus the up arrow to perform an action.

**Comment by titet11** [comment]
@XLTechie ok, I will keep it in mind, although I do not speak English and therefore, I forget to point out that section

**Comment by titet11** [comment]
@XLTechie 
Although my answers may be misinterpreted, anyway so far you have all managed to understand me, and that is the important thing. I think this topic is already understood, or not.
